File: python-kata/generator.py
#!/usr/bin/env python3

import json
import os
import shutil
from pathlib import Path
from algorithm_definitions import get_definitions

def get_next_day_number():
    """Calculate the next day number based on existing day directories."""
    kata_dir = Path("kata")
    if not kata_dir.exists():
        return 1
    
    day_numbers = []
    for item in kata_dir.iterdir():
        if item.is_dir() and item.name.startswith("day"):
            try:
                day_num = int(item.name[3:])
                day_numbers.append(day_num)
            except ValueError:
                continue
    
    return max(day_numbers, default=0) + 1

def generate_class(day_path, class_name, definition):
    """Generate a Python class file."""
    lines = []
    
    # Imports
    lines.append("from typing import Optional, List, TypeVar")
    lines.append("from kata.types import *")
    lines.append("from kata.interfaces import List as ListInterface")
    lines.append("")
    
    # Generic type variable if needed
    if definition.generic:
        generic_clean = definition.generic.replace("[", "").replace("]", "")
        lines.append(f"{generic_clean} = TypeVar('{generic_clean}')")
        lines.append("")
    
    # Class definition
    class_line = f"class {class_name}"
    if definition.generic:
        class_line += f"(ListInterface{definition.generic})" if "List" in class_name else ""
    class_line += ":"
    lines.append(class_line)
    
    # Properties
    for prop in definition.properties:
        lines.append(f"    def __init__(self):")
        lines.append(f"        self.{prop.name}: {prop.type} = 0")
        lines.append("")
        break  # Only need __init__ once
    
    if not definition.properties:
        lines.append("    def __init__(self):")
        lines.append("        pass")
        lines.append("")
    
    # Methods
    for method in definition.methods:
        method_line = f"    def {method.name}(self"
        if method.parameters:
            method_line += f", {method.parameters}"
        method_line += f") -> {method.return_type}:"
        lines.append(method_line)
        
        # Default return values
        if method.return_type == "None":
            lines.append("        pass")
        elif method.return_type == "bool":
            lines.append("        return False")
        elif method.return_type == "int":
            lines.append("        return 0")
        elif "List" in method.return_type:
            lines.append("        return []")
        else:
            lines.append("        return None")
        lines.append("")
    
    # Write file
    file_path = day_path / f"{class_name}.py"
    with open(file_path, 'w') as f:
        f.write('\n'.join(lines))

def generate_function(day_path, class_name, definition):
    """Generate a Python function file."""
    lines = []
    
    # Imports
    lines.append("from typing import Optional, List")
    lines.append("from kata.types import *")
    lines.append("")
    
    # Function definition
    func_line = f"def {definition.function_name}({definition.parameters}) -> {definition.return_type}:"
    lines.append(func_line)
    
    # Default return values
    if definition.return_type == "None":
        lines.append("    pass")
    elif definition.return_type == "bool":
        lines.append("    return False")
    elif definition.return_type == "int":
        lines.append("    return -1")
    elif "List" in definition.return_type:
        lines.append("    return []")
    else:
        lines.append("    return None")
    
    # Write file
    file_path = day_path / f"{class_name}.py"
    with open(file_path, 'w') as f:
        f.write('\n'.join(lines))

def update_stats(dsa_list):
    """Update statistics file."""
    stats_file = Path("stats.json")
    stats = {}
    
    if stats_file.exists():
        with open(stats_file) as f:
            stats = json.load(f)
    
    # Update stats
    for ds in dsa_list:
        stats[ds] = stats.get(ds, 0) + 1
    
    with open(stats_file, 'w') as f:
        json.dump(stats, f, indent=2)

def main():
    # Read configuration
    with open("kata_config.json") as f:
        config = json.load(f)
    
    dsa_list = config["dsa"]
    
    # Calculate next day number
    day = get_next_day_number()
    day_name = f"day{day}"
    day_path = Path("kata") / day_name
    
    # Clean and create day directory
    if day_path.exists():
        shutil.rmtree(day_path)
    day_path.mkdir(parents=True)
    
    # Create __init__.py for the day package
    (day_path / "__init__.py").touch()
    
    # Get algorithm definitions
    definitions = get_definitions()
    
    # Generate files for each algorithm
    for ds_name in dsa_list:
        definition = definitions.get(ds_name)
        if not definition:
            raise RuntimeError(f"Algorithm {ds_name} not found in definitions")
        
        if definition.type == "class":
            generate_class(day_path, ds_name, definition)
        else:
            generate_function(day_path, ds_name, definition)
    
    # Update statistics
    update_stats(dsa_list)
    
    print(f"Generated {day_name} with {len(dsa_list)} algorithms")

if __name__ == "__main__":
    main()

File: java-kata/src/main/java/com/kata/Generator.java
package com.kata;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.kata.AlgorithmDefinitions.*;

import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.stream.Collectors;

public class Generator {
    private static final String SRC_PATH = "src/main/java/com/kata";
    private static final String TEST_PATH = "src/test/java/com/kata";
    
    public static void main(String[] args) throws IOException {
        generate();
    }
    
    public static void generate() throws IOException {
        // Read configuration
        ObjectMapper mapper = new ObjectMapper();
        JsonNode config = mapper.readTree(new File("kata.config.json"));
        List<String> dsaList = new ArrayList<>();
        config.get("dsa").forEach(node -> dsaList.add(node.asText()));
        
        // Calculate next day number
        int day = getNextDayNumber();
        String dayName = "day" + day;
        Path dayPath = Paths.get(SRC_PATH, dayName);
        
        // Clean and create day directory
        if (Files.exists(dayPath)) {
            deleteDirectory(dayPath);
        }
        Files.createDirectories(dayPath);
        
        // Get algorithm definitions
        Map<String, AlgorithmDef> definitions = AlgorithmDefinitions.getDefinitions();
        
        // Generate files for each algorithm
        for (String dsName : dsaList) {
            AlgorithmDef def = definitions.get(dsName);
            if (def == null) {
                throw new RuntimeException("Algorithm " + dsName + " not found in definitions");
            }
            
            if ("class".equals(def.type)) {
                generateClass(dayPath, dsName, def);
            } else {
                generateFunction(dayPath, dsName, def);
            }
        }
        
        // Update configuration files
        updatePomXml(dayName);
        updateStats(dsaList);
        
        System.out.println("Generated " + dayName + " with " + dsaList.size() + " algorithms");
    }
    
    private static int getNextDayNumber() throws IOException {
        Path srcPath = Paths.get(SRC_PATH);
        if (!Files.exists(srcPath)) {
            return 1;
        }
        
        try {
            return Files.list(srcPath)
                    .filter(Files::isDirectory)
                    .map(path -> path.getFileName().toString())
                    .filter(name -> name.startsWith("day"))
                    .mapToInt(name -> {
                        try {
                            return Integer.parseInt(name.substring(3));
                        } catch (NumberFormatException e) {
                            return 0;
                        }
                    })
                    .max()
                    .orElse(0) + 1;
        } catch (IOException e) {
            return 1;
        }
    }
    
    private static void generateClass(Path dayPath, String className, AlgorithmDef def) throws IOException {
        StringBuilder sb = new StringBuilder();
        sb.append("package com.kata.").append(dayPath.getFileName()).append(";\n\n");
        
        // Add imports
        sb.append("import com.kata.types.*;\n");
        sb.append("import com.kata.interfaces.*;\n");
        sb.append("import java.util.*;\n\n");
        
        // Class declaration
        sb.append("public class ").append(className);
        if (def.generic != null) {
            sb.append(def.generic);
        }
        sb.append(" {\n");
        
        // Properties
        for (Property prop : def.properties) {
            sb.append("    ").append(prop.scope).append(" ").append(prop.type).append(" ").append(prop.name).append(";\n");
        }
        
        if (!def.properties.isEmpty()) {
            sb.append("\n");
        }
        
        // Constructor
        sb.append("    public ").append(className).append("() {\n");
        sb.append("        \n");
        sb.append("    }\n");
        
        if (!def.methods.isEmpty()) {
            sb.append("\n");
        }
        
        // Methods
        for (Method method : def.methods) {
            sb.append("    public ").append(method.returnType).append(" ").append(method.name).append("(").append(method.parameters).append(") {\n");
            if (!"void".equals(method.returnType)) {
                if ("int".equals(method.returnType)) {
                    sb.append("        return 0;\n");
                } else if ("boolean".equals(method.returnType)) {
                    sb.append("        return false;\n");
                } else {
                    sb.append("        return null;\n");
                }
            }
            sb.append("    }\n\n");
        }
        
        sb.append("}\n");
        
        Files.write(dayPath.resolve(className + ".java"), sb.toString().getBytes());
    }
    
    private static void generateFunction(Path dayPath, String className, AlgorithmDef def) throws IOException {
        StringBuilder sb = new StringBuilder();
        sb.append("package com.kata.").append(dayPath.getFileName()).append(";\n\n");
        
        // Add imports
        sb.append("import com.kata.types.*;\n");
        sb.append("import java.util.*;\n\n");
        
        // Class with static method
        sb.append("public class ").append(className).append(" {\n");
        sb.append("    public static ").append(def.returnType).append(" ").append(def.functionName).append("(").append(def.parameters).append(") {\n");
        if (!"void".equals(def.returnType)) {
            if ("int".equals(def.returnType)) {
                sb.append("        return 0;\n");
            } else if ("boolean".equals(def.returnType)) {
                sb.append("        return false;\n");
            } else {
                sb.append("        return null;\n");
            }
        }
        sb.append("    }\n");
        sb.append("}\n");
        
        Files.write(dayPath.resolve(className + ".java"), sb.toString().getBytes());
    }
    
    private static void updatePomXml(String dayName) throws IOException {
        // Read current pom.xml
        String pomContent = Files.readString(Paths.get("pom.xml"));
        
        // Update current.day property
        pomContent = pomContent.replaceAll(
                "<current\\.day>day\\d+</current\\.day>",
                "<current.day>" + dayName + "</current.day>"
        );
        
        Files.write(Paths.get("pom.xml"), pomContent.getBytes());
    }
    
    private static void updateStats(List<String> dsaList) throws IOException {
        ObjectMapper mapper = new ObjectMapper();
        Map<String, Integer> stats = new HashMap<>();
        
        File statsFile = new File("stats.json");
        if (statsFile.exists()) {
            JsonNode existingStats = mapper.readTree(statsFile);
            existingStats.fieldNames().forEachRemaining(fieldName ->
                stats.put(fieldName, existingStats.get(fieldName).asInt())
            );
        }
        
        // Update stats
        for (String ds : dsaList) {
            stats.put(ds, stats.getOrDefault(ds, 0) + 1);
        }
        
        mapper.writerWithDefaultPrettyPrinter().writeValue(statsFile, stats);
    }
    
    private static void deleteDirectory(Path path) throws IOException {
        if (Files.exists(path)) {
            Files.walk(path)
                    .sorted(Comparator.reverseOrder())
                    .map(Path::toFile)
                    .forEach(File::delete);
        }
    }
}